Use async/await, cleanup fetch() code

josephfrazier opened the pull request:
This started out as an effort to use async/await instead of then/catch/callbacks, and as I went along, I realized that the `fetch` function could be called from the content script rather than the background script (see [MDN](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Content_scripts#Cross-domain_privileges)) and we could get rid of all the message passing wrappers around `fetch`, so I did that too. I also made `fetch` iterative rather than recursive.

Interestingly, even though using async/await requires babel to insert extra runtime code, the total extension size is smaller (probably because the background script no longer includes jquery):

    du -shc dist/*js

I tried to make each commit sensible on its own (some might be easier to view while ignoring whitespace), and to have the overall commit history follow my thought processes. Let me know if you have any questions/concerns!

stefanbuck commented:
I added the background script while I was working on adding Firefox support #100. At this time it was the only way to do requests. Great to see they getting closer to chrome extension. I need a bit more time to review this.

josephfrazier commented:
Oh, I see now why the background requests were necessary. Also glad that it has changed since then.

Take your time reviewing, I know there are some big changes here (and I was a little low on sleep when making them). I'll add some inline comments about parts that might need extra care.

josephfrazier commented on lib/click-handler.js:
```diff
@@ -110,19 +90,38 @@ function onClick(event) {
+    chrome.runtime.sendMessage({
+      type: 'fetch',
+      payload: urls,
+    });
```
As mentioned in the comment, the tests currently stub out `sendMessage` and rely on it to be called. I briefly tried using https://github.com/speedskater/babel-plugin-rewire to do the same thing for our `fetch` module, but I wasn't able to get it working, so I just left things as they are for now.

stefanbuck replied:
In this case I would remove the code from here (since it isn't used anymore) and disable the related test until we find a suitable solution

josephfrazier replied:
Agreed, fixed in https://github.com/OctoLinker/browser-extension/pull/306/commits/af77eda38aab5c0aaa186c07db2261cd49f2345e

josephfrazier commented on test/click-handler.test.js:
```diff
@@ -10,6 +10,7 @@ describe('click-handler', () => {
   let resolveFake;
 
   sandbox.stub(window.chrome.runtime, 'sendMessage');
+  window.console.error = () => {}; // Suppress fake onClick errors
```
Before, the `return console.error(err);` line in `onClick` wasn't getting run because `sendMessage` was stubbed, so `tryLoad` never called its callback. Now, `console.error` gets called because fetch is an async function, so we just need to silence it here (to avoid having the tests print useless output). See my other comment for details about the stubbing: https://github.com/OctoLinker/browser-extension/pull/306/files#r104506996).

josephfrazier commented on package.json:
```diff
@@ -28,6 +28,12 @@
+      [
+        "transform-runtime",
+        {
+          "polyfill": false
```
`babel-plugin-transform-runtime` is used instead of `babel-polyfill` due to
https://github.com/OctoLinker/browser-extension/pull/258#issuecomment-275813737

Note one of them is required. `babel-preset-env` alone isn't enough:
https://github.com/babel/babel/issues/5085#issuecomment-271640887

josephfrazier commented on lib/options/storage.js:
```diff
@@ -12,43 +12,44 @@ options.forEach((item) => {
+  try {
+    return await chromep.storage.sync.set(data);
+  } catch (err) {
+    return chromep.storage.local.set(data);
```
Yes. I had originally put `await` in there, but eslint complained:

    /Users/josephfrazier/workspace/octolinker/browser-extension/lib/options/storage.js
      23:12  error  Redundant use of `await` on a return value  no-return-await

This makes sense because `async` functions always return a promise, so you don't have to `await` the return value. The reason that `await` is used inside the `try` block is so that any errors with `chromep.storage.sync.set(data)` will be caught inside this function, rather than being exposed to the function caller.

josephfrazier commented on lib/background/fetch.js:
```diff
@@ -1,48 +1,23 @@
-          category: 'fetch',
-          action: 'error',
-        },
+export default async (urls) => {
```
Good point, fixed in https://github.com/OctoLinker/browser-extension/pull/306/commits/b896d9dea3d43eb4cf405337c34ea15104e94feb

josephfrazier commented:
Thanks for the review! I addressed the comments and merged this in. I also created https://github.com/OctoLinker/browser-extension/issues/311 for fixing the click-handler/fetch tests.
